Model system rework

gigaherz opened the pull request:
This is the initial implementation of my new model system idea.

This feature expands the vanilla model loading with a new key: "loader". If this key is found in the model, forge will derive a model instance from any necessary values seen in the json.

Example, `models/item/test_fluid_bucket.json` 
```json
{
  "parent": "forge:item/bucket",
  "loader": "forge:bucket",
  "fluid": "new_fluid_test:test_fluid"
}
```

The "parent" key is handled by the vanilla logic, which in this case takes care of providing the required default textures for the bucket. The "forge:bucket" loader constructs a custom model that uses a base texture and some mask textures, to produce a custom bucket model for the given fluid in the "fluid" key.

### I need people to test this! Here's how:

1. Clone/checkout the forge repository `git clone https://github.com/MinecraftForge/MinecraftForge.git`
2. Fetch and checkout my branch `git fetch origin pull/6180/head:model-system-rework` + `git checkout model-system-rework`
3. Run `gradlew setup`
4. Run `gradlew publishToMavenLocal`
5. Go to `<forge folder>/projects/forge/build/libs/` and write down the exact version number eg. `forge-1.14.4-28.1.90-model-system-rework.jar`
6. Edit your project's build.gradle, and in your repositories (NOT the repositories block from `buildscript`!!), add `mavenLocal()` as a repository, eg.
```gradle
repositories {
    mavenLocal()
}
```
7. Edit your minecraft dependency to match the version as seen in the build results, eg.
```gradle
minecraft 'net.minecraftforge:forge:1.14.4-28.1.90-model-system-rework'
```
8. Refresh/reimport your gradle project into your IDE, and wait for the setup to complete.
9. Rerun `gradlew gen<YOURIDE>Runs`
10. Implement your models under the new system and tell me of any issue you encounter or any feature you think would be important to add.

gigaherz commented:
Showcase:

![test fluid bucket](https://user-images.githubusercontent.com/685851/65395307-6c67fd80-dd99-11e9-9f26-3f85e5765c37.png)

This is the test fluid bucket, which I gave a model json that uses the new system.

gigaherz commented:
Status so far:
  * Modified BlockModel to redirect parsing model files to the new loader ✔️ 
    * Expanded BlockModel to hold the (optional) custom model ✔️ 
      * Model jsons now support a "loader" key as an alternative to the vanilla "elements" key. ✔️ 
    * Expanded BlockModel to hold the (optional) forge-style transforms data ✔️ 
      * Model jsons now support a "transform" key as an alternative to the vanilla "display" key. ✔️ 
  * Implemented loader support for using existing old-style model loaders from model jsons ✔️ 
  * Implemented brand new OBJ model loader designed for the new system ✔️ 
  * Adapted existing ModelDynBucket to handle the new model system API ✔️ 
  * Tested that the new model loader system does not prevent vanilla models from loading correctly ✔️ 
  * Tested that the new model loader system does not prevent forge blockstates models from loading correctly ✔️

williewillus commented:
I guess removing all the *2 classes isnt possible without breaking compat, in which case you volunteer to clean it up next major version ;P

Otherwise, this looks great from a user interface point of view. Could you maybe give a brief high-level rundown of what code changes were done?

gigaherz commented:
Yes the goal is to get rid of the old model loading system in 1.15, and rename those *2 classes.

EDIT: Err premature sending.

As for code changes... Note that the PR has more changes unrelated to the model system because #6154 has not been merged yet. This includes changes to BlockModelDefinition, ModelBakery, Variant, BlockStateLoader, ModelLoader, ...

I have outlined the changes to BlockModel in the comment above, which is where most of the hooks reside in.
I have expanded IForgeBakedModel to include a `doesHandlePerspective` getter, which needs to be overriden to `true` in order to avoid the extra wrapping of the model into a `PerspectiveMapWrapper`.
I have moved the TRSR transform deserializer logic to the new model loader class, so that it doesn't get lost when cleaning up next version.
I have provided actual PNG textures for forge:white (used to be a custom sprite), masks for the new vanilla bucket icon.

williewillus commented:
Are there plans to add some common model types similar to the old forge ones?

One I can think of already is just composing two other models together with individual transforms for each

gigaherz commented:
I didn't have any plans for it, but that doesn't mean it can't be done. :P

gigaherz commented:
Rebased on top of latest forge, now that I merged #6154

This screenshot shows:
 - a working bucket model
 - the same model displaying increased size from the forge-style transforms
 - blockstates-based "obsidian wall" (not a fully working block, only a test)
 - OBJ model in inventory and head.

![image](https://user-images.githubusercontent.com/685851/66353957-a9b8b780-e963-11e9-947f-6fac7c5dba0a.png)

Alpvax commented:
Is it just me, or is that bucket huge? If so, is that intentional?

TeamSpen210 commented:
That's the default size of a model, without the item transformation settings. That wouldn't be important though.

Alpvax commented:
My question was is there an error with the size, the answer is no. Thanks

gigaherz commented:
I have updated the top description to include steps to help people test.

KlemenDEV commented:
I have tested your branch and found out the following:

- the old way to load OBJ models still works
- vanilla model loading is not broken
- the new bucket JSON format works
- new way of OBJ loading did not work for me (tested with item model)

I have tried to load OBJ model for item. The item model JSON code is:

```json
{
  "parent": "forge:item/default",
  "loader": "forge:obj",
  "model": "test:models/item/item_direction.obj",
  "textures": {
    "particle": "test:items/particle.png"
  },
  "transform": "forge:default-item"
}
```


The error I see in the client run log is:

`Unable to load model: 'test:obj#inventory' referenced from: test:obj#inventory: java.lang.RuntimeException: Could not read OBJ model`

The OBJ and MTL files are attached in the ZIP. OBJ file is loaded properly with the old OBJ loader.

[item_direction.zip](https://github.com/MinecraftForge/MinecraftForge/files/3847665/item_direction.zip)

KlemenDEV commented:
Now OBJ model works just fine, and other things I have tested and listed above too.

AzureDoom commented:
Works for great for OBJ's that just use the MTL to get the textures and even get the multiple textures from the MTL. No code needed, just use the `"loader": "forge:obj",` and set it up correctly. Just have to update all of my transforms from 1.12. 596 jsons more to go! 
![image](https://user-images.githubusercontent.com/7415711/68923026-47fe2100-074b-11ea-89f6-c4f42a8ff970.png)

nikita488 commented:
It is possible to add custom json attribute to OBJModel that will specify group that will be renderer and not all groups from obj file? Like for example i have Crystal obj file with 6 different groups (based on amount of crystals in block) and i want to render just one of them, not all. Without this i need to split my obj file to 6 files.

gigaherz commented:
I **really** want to avoid coding logic for that. It makes everything a huge mess for little gain. 
Your specific use case however does have some merit.

I set up the system so that part information is available internally, for loaders that produce multi-part models. I suppose I could maybe set up a "part predicate" system, that any loader can make use of. I'll look into it and see if I can come up with a part whitelist/blacklist system that isn't a mess.

gigaherz commented:
@nikita488 I hope this satisfies your needs. :P

KlemenDEV commented:
I have tested the current branch after rebase (1.14.4-28.1.116) and all seems to be working fine. I hope this one gets reviewed soon so we can get these features in a version that can be downloaded by the mod users :)

gigaherz commented:
Yep, I think the base system is already quite robust. Big missing feature right now is expanding the model data generators to make them loader-aware.
 
This dynamic model stuff I'm doing right now will probably be stripped away into a followup PR, so I don't see it as blocking the review.

desht commented:
I'm having big problems getting this to load OBJ models.  Here's a PneumaticCraft Air Compressor, which loads correctly using the default 1.14.4 loader and a blockstates file like https://github.com/TeamPneumatic/pnc-repressurized/blob/master/src/main/resources/assets/pneumaticcraft/blockstates/air_compressor.json   - minus the "inventory" section, that file's from 1.12):
![2019-11-18_10 59 12](https://user-images.githubusercontent.com/736831/69053535-87559780-0a01-11ea-8a86-5aea9a32a363.png)

(and ignore the corner rendering, separate issue)

Here's the reworked blockstate file:
* https://github.com/TeamPneumatic/pnc-repressurized/blob/1.14/src/main/resources/assets/pneumaticcraft/blockstates/air_compressor.json

New JSON block models:
* https://github.com/TeamPneumatic/pnc-repressurized/blob/1.14/src/main/resources/assets/pneumaticcraft/models/block/air_compressor_off.json
* https://github.com/TeamPneumatic/pnc-repressurized/blob/1.14/src/main/resources/assets/pneumaticcraft/models/block/air_compressor_on.json

OBJ and MTL files (unchanged):
* https://github.com/TeamPneumatic/pnc-repressurized/blob/1.14/src/main/resources/assets/pneumaticcraft/models/block/air_compressor.obj
* https://github.com/TeamPneumatic/pnc-repressurized/blob/1.14/src/main/resources/assets/pneumaticcraft/models/block/air_compressor.mtl
* https://github.com/TeamPneumatic/pnc-repressurized/blob/1.14/src/main/resources/assets/pneumaticcraft/models/block/air_compressor_active.obj
* https://github.com/TeamPneumatic/pnc-repressurized/blob/1.14/src/main/resources/assets/pneumaticcraft/models/block/air_compressor_active.mtl

New item JSON:
* https://github.com/TeamPneumatic/pnc-repressurized/blob/1.14/src/main/resources/assets/pneumaticcraft/models/item/air_compressor.json

And here's the result:

![2019-11-18_12 44 10](https://user-images.githubusercontent.com/736831/69053588-a94f1a00-0a01-11ea-8f45-0ad1af08350e.png)

There are 4 rotations there: north, south, west, east from left to right.  But the model hasn't been rotated at all and the textures are wrongly applied.

And here's what happens when the "on" block property goes to true:

![2019-11-18_12 33 26](https://user-images.githubusercontent.com/736831/69053702-e9160180-0a01-11ea-8f85-11750d7ccd75.png)

A completely black texture.

Finally the item model is... well, isn't.  The item is completely invisible, whether held or on the ground.

There are no errors pertaining to the model or textures in the log file.

Am I doing something completely wrong here?

desht commented:
Just updated to the latest branch build (1.14.4-28.1.120-model-system-rework) and it's looking very much better now... but, one little problem remains that I can see: when I place down the air compressor facing east or west all is well, but if facing north or south, the model appears to be rotated 180°.

Of course, that's trivially fixed by changing the rotations in the blockstates file, but it does appear to rotate in the opposite direction from how it used to...

gigaherz commented:
Huh, I use the existing logic for applying transforms.... this would have to be a mismatch between the rotation direction (clockwise vs counter-clockwise) between blockstates jsons, and the forge "transforms" logic? will investigate. but worst case, we may just have to deal with it.

desht commented:
Yeah, I can deal with it, no worries.  Other than that, it appears to be working great, so I really hope this makes it in.  I'll be converting all the other PNC models over in the next few days and will get some more testing done, but for now:
![2019-11-20_00 57 47](https://user-images.githubusercontent.com/736831/69199790-01d10500-0b31-11ea-8e8c-8b92ecf544a7.png)

desht commented:
Regarding the rotation issue, after a bit more research, I think the way you do it now is arguably more correct than before, as in more consistent with vanilla.  E.g. https://minecraft.gamepedia.com/Model#Example:_Wall_Torch is doing a clockwise rotation for positive "y" values.  So I'd say leave it as is... tweaking a few rotations in blockstates files is a simple one-off change.

desht commented:
I noticed that the `forge:bucket` loader doesn't appear to account for fluid colours, as defined by `FluidAttributes.Builder#color()`.  This attribute *does* affect the rendering of the fluid in-world, but the bucket item always renders the fluid texture exactly as it was loaded.

gigaherz commented:
Oops forgot to reply earlier, I wasn't home. The bucket model has a "applyTint" value, which I mistakenly made default false. I'll fix.

desht commented:
Just a quick comment to say that tinted OBJ models appear to work very nicely:
![2019-11-28_23 57 11](https://user-images.githubusercontent.com/736831/69835745-962b1e00-123b-11ea-8e4f-bdace5f1ce38.png)
The red & blue sections on that Vortex Tube are OBJ-tinted and vary depending on the temperature of each end of the tube.  Works great!

KlemenDEV commented:
@gigaherz Are there any updates on when this feature branch could be merged with master 1.14.4 branch so these features could be used in production mods?

gigaherz commented:
Considering it's 3-4 days until the release of 1.15, this will most probably be merged straight into the 1.15 branch, and I'll take care of any remaining polish/bugs during the 1.15 testing phase. Whether or not I'll try to also get it into 1.14, depends on if 1.14.4 becomes the next LTS version, or 1.12.2 remains LTS until 1.16

AzureDoom commented:
Just wondering if there is any update to getting this merged into 1.14.4, I know you said about 1.15 but as we don't know how long that may be before it releases, 1.14.4 mods would greatly benefit from this system. I have myself have built my mods against the features here since they are greatly preferred to the current system, where we'd have to register each OBJ in code to get it work and then do the transforms in code. As someone with 500 OBJ's, your system is much easier.

gigaherz commented:
We are all busy getting 1.15 up and running. This pr is on hold for a few days.

gigaherz commented:
With this last commit, I should have brought this PR in line with the 1.15 changes. Everything seems to work locally, but if I could get a thorough testing AND code review of this, from anyone who has the time, to ensure I didn't introduce any last-minute WTFs, it would be greatly appreciated. I hope to be able to merge this within the next few days, now that 1.14.4 is the next LTS version.

Note that I squished the commits into one (now two), so make sure you test the right build (if you already had a build with that name, you may need to wipe the gradle & mavenLocal caches, or change the local branch name so that the files generate with a different suffix)

desht commented:
Although I haven't done a code review, I can confirm that all my in-dev PneumaticCraft models (both item and block) continue to work fine with the latest commit.

(Oh, and the bucket model now properly tints fluids, so yay!)

AzureDoom commented:
I can't confirm all other extra things you support now but in my use case of 600+ models, each one loads fine and works as intended perfectly.

KlemenDEV commented:
I have tested all the features I mentioned in [my older comment](https://github.com/MinecraftForge/MinecraftForge/pull/6180#issuecomment-553978887) and all of these features work fine. I would like to thank to @gigaherz and all other people who worked on this. I hope we can get this in 1.14.4 soon :)

tterrag1098 commented on patches/minecraft/net/minecraft/client/renderer/texture/TextureAtlasSprite.java.patch:
```diff
@@ -42,14 +42,27 @@
+    }
+ 
+    private int func_195665_a(int p_195665_1_, int p_195665_2_, int p_195665_3_, int p_195665_4_) {
++      if (this.field_195671_d == null || this.field_195672_e == null) return this.field_195670_c[p_195665_2_].func_195709_a(p_195665_3_, p_195665_4_);
```
Could use some comments here describing this change. I'm not totally sure what it's for just looking at it.

tterrag1098 commented on src/main/java/net/minecraftforge/client/model/CompositeModel.java:
```diff
@@ -0,0 +1,272 @@
+        List<BakedQuad> quads = new ArrayList<>();
+        for(Map.Entry<String, IBakedModel> entry : bakedParts.entrySet())
+        {
+            // TODO: Some way to provide submodel data?
```
What's the plan for this TODO?

gigaherz replied:
Looking at my code again (yes it has been long enough that I wasn't certain how I wrote that), seems I did provide MOST of what's needed for a submodel data API. I just didn't provide an explicit way to access it.
I have refactored that part of the code into what I hope is the best way to achieve this purpose

tterrag1098 commented on src/main/java/net/minecraftforge/client/model/CompositeModel.java:
```diff
@@ -0,0 +1,272 @@
+            ImmutableMap.Builder<String, Submodel> parts = ImmutableMap.builder();
+            for(Map.Entry<String, JsonElement> part : modelContents.get("parts").getAsJsonObject().entrySet())
+            {
+                // TODO: Allow customizing state? If so, how?
```
Could at least add x/y rotation fields ala vanilla variants?

gigaherz replied:
Ah I should get rid of this comment. The "how" answered itself a while ago through the "transform" feature.
You can do something like this:
```json
{
"loader": "forge:composite",
"parts": {
  "part1: {
    "parent": x,
    "transform": { "rotation": ..., "translation": ..., "scale" }
  }
}
```

tterrag1098 replied:
It might still be worthwhile to support the much simpler vanilla-style x/y rotation, since that can be trivially converted to a TRSR by existing code.

gigaherz replied:
That would mean adding even more values to the model deserializer, or creating a dedicated deserializer for this. It's currently parsed as if it was any other model json file.

tterrag1098 commented on src/main/java/net/minecraftforge/client/model/DummyAtlasSprite.java:
```diff
@@ -0,0 +1,38 @@
+
+import java.util.function.Function;
+
+public class DummyAtlasSprite extends TextureAtlasSprite
```
Where is this used?

gigaherz replied:
Oops leftover from the "dynamic models" stuff I scrapped.

tterrag1098 commented on src/main/java/net/minecraftforge/client/model/ModelLoaderRegistry2.java:
```diff
@@ -0,0 +1,507 @@
+    // Forge built-in loaders
+    public static void init()
+    {
+        registerLoader(new ResourceLocation("forge:obj"), OBJLoader2.INSTANCE);
```
These should use the separated constructor, not hardcode the colon.

tterrag1098 commented on src/main/java/net/minecraftforge/client/model/ModelLoaderRegistry2.java:
```diff
@@ -0,0 +1,507 @@
+        try
+        {
+            if (!loaders.containsKey(loaderId))
+                throw new IllegalStateException(String.format("Model loader '%s' not found.", loaderId));
```
Perhaps dump a list of available loaders here

tterrag1098 commented on src/main/java/net/minecraftforge/client/model/ModelLoaderRegistry2.java:
```diff
@@ -0,0 +1,507 @@
+        return getModel(loader, deserializationContext, object);
+    }
+
+    private static final Pattern FILESYSTEM_PATH_TO_RESLOC = Pattern.compile("(?:.*[\\\\/]assets[\\\\/](?<namespace>[a-z_-]+)[\\\\/]textures[\\\\/])?(?<path>[a-z_\\\\/-]+)\\.png");
```
Could you document this behemoth of a regex?

tterrag1098 commented on src/main/java/net/minecraftforge/client/model/ModelLoaderRegistry2.java:
```diff
@@ -0,0 +1,507 @@
+    public static ResourceLocation resolveTexture(@Nullable String tex, IModelConfiguration owner)
+    {
+        if (tex == null)
+            return new ResourceLocation("forge:white");
```
Shouldn't this be a constant somewhere

tterrag1098 commented on src/main/java/net/minecraftforge/client/model/ModelLoaderRegistry2.java:
```diff
@@ -0,0 +1,507 @@
+
+            try
+            {
+                IUnbakedModel model = parent.loadModel(new ResourceLocation(modelLocation));
```
This doesn't check `accepts` anywhere which could cause strange problems

gigaherz replied:
I assumed the person choosing the loader would know what loader to use, but on second thought...

tterrag1098 commented on src/main/java/net/minecraftforge/client/model/ModelLoaderRegistry2.java:
```diff
@@ -0,0 +1,507 @@
+                model.customData.setCustomGeometry(geometry);
+            }
+
+            IModelState modelState = deserializeModelTransforms(deserializationContext, jsonobject);
```
Instead of calling these manually, shouldn't these types get their own type adapters and then use `context.deserialize` here?

gigaherz replied:
This method does more than just read a TRSRTransformation object, it also handles things like the stock entries such as "forge:default-item". I might be able to change the logic in 1.15, but in this PR it doesn't make sense as this is a shared method with the old loader and I'd rather not do major refactoring on the old loader.

tterrag1098 commented on src/test/java/net/minecraftforge/debug/client/model/NewModelLoaderTest.java:
```diff
@@ -0,0 +1,102 @@
+            new Item(new Item.Properties().group(ItemGroup.MISC))
+    );
+
+    public NewModelLoaderTest()
```
Shouldn't this also test adding a custom loader?

gigaherz replied:
Possibly, yeah. Actually just looking at it brought something to my attention: we might run into race conditions if people call registerLoader from the FMLClientSetupEvent. I'll make the loader registry a concurrent map while at it.

gigaherz committed:
Backport implementation of the new model system from 1.15:
- Extended block model format to support a "loader" value, as an alternative to "elements.
- Extended block model format to support "gui_3d":false as a way to mark a model as being intended to be flat and lit from the front.
    (No this is not equivalent to the  "gui_light" flag in 1.15.2. It changes more than just the light direction)
- Added a new OBJ loader designed from the ground up to work with the new model system.
- Added stock models to serve as parent models when using custom loaders (forge:item/default, forge:item/default-tool, and forge:block/default).
- Added stock white texture (forge:white) for use in loaders that want to draw untextured geometry.
- Deprecated the old loader components, since they are gone in 1.15 already.

gigaherz commented:
Welp, this is it.
Rechecked 4 times. Everything works fine as far as I can tell.
Here goes nothing.
